# Verificando a perfomace dos modelo sem otimização

In [21]:
# Importando as blibiotecas

import numpy as np 
import pandas as pd 

from sklearn.svm import SVC 
from sklearn.naive_bayes import GaussianNB 
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import cross_validate, KFold
from imblearn.under_sampling import RandomUnderSampler
from sklearn.linear_model import LogisticRegression


In [3]:
breast = pd.read_csv('data.csv')

In [4]:
breast.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [5]:
breast.drop(['id','Unnamed: 32'],axis=1,inplace=True)

In [6]:
breast.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

In [7]:
breast['diagnosis'] = breast['diagnosis'].map({'M':1,'B':0})

In [8]:
breast.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1        17.99         10.38          122.80     1001.0   
1          1        20.57         17.77          132.90     1326.0   
2          1        19.69         21.25          130.00     1203.0   
3          1        11.42         20.38           77.58      386.1   
4          1        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

In [9]:
# Separando os dados e atributos e target
x = breast.drop(['diagnosis'],axis=1).values
y = breast.diagnosis.values

# Normalizando os dados
x_std = StandardScaler().fit_transform(x)

In [10]:
# usando o metos undersampling para tornar os dados menos enviesados e comparar as metricas
nm = RandomUnderSampler(random_state = 10)

x_nm, y_nm = nm.fit_resample(x_std, y)

In [22]:
svm = SVC()
knn = KNeighborsClassifier()
tree = DecisionTreeClassifier()
na_ba = GaussianNB()
lg = LogisticRegression()


In [12]:
# colocando o kfold
kfold = KFold(n_splits=7)

#### Aplicando o svm

In [13]:
nome_metricas = ['accuracy', 'precision_macro', 'recall_macro']

metricas_ran = cross_validate(svm,x_std, y, cv=kfold, scoring=nome_metricas)

for met in metricas_ran:
  print(f'-{met}')
  print(f"-- {metricas_ran[met]}")
  media = np.mean(metricas_ran[met])
  desvio = np.std(metricas_ran[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.09580278 0.01180577 0.01112485 0.01131964 0.01061845 0.01113629
 0.01095605]
Média do fit_time: 0.02325197628566197
Desvio 0.029620622214247206
Intervalo [-0.036,0.082]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.06403518 0.00591397 0.0049088  0.00445318 0.00536227 0.00454402
 0.00528836]
Média do score_time: 0.013500826699393136
Desvio 0.020635852138259168
Intervalo [-0.028,0.055]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.96341463 0.95121951 0.98765432 0.96296296 1.         0.98765432
 0.96296296]
Média do test_accuracy: 0.9736955306060996
Desvio 0.01657392539064597
Intervalo [0.941,1.007]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.95387205 0.9474359  0.98863636 0.97368421 1.         0.97222222
 0.94607088]
Média do test_precision_macro: 0.9688459465667316
Desvio 0.01926638482843045
Intervalo [0.930,1.007]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*

In [14]:
metricas_ran_u = cross_validate(svm,x_nm, y_nm, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_u:
  print(f'-{met}')
  print(f"-- {metricas_ran_u[met]}")
  media = np.mean(metricas_ran_u[met])
  desvio = np.std(metricas_ran_u[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.01703358 0.01614046 0.03977203 0.0121789  0.00874352 0.00931454
 0.00815797]
Média do fit_time: 0.01590585708618164
Desvio 0.010278866574167739
Intervalo [-0.005,0.036]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.02039337 0.0668478  0.01544285 0.01280236 0.0145328  0.01610541
 0.00762224]
Média do score_time: 0.021963834762573242
Desvio 0.0186651517706428
Intervalo [-0.015,0.059]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [1.         0.95081967 0.98360656 0.96721311 0.91666667 0.95
 0.98333333]
Média do test_accuracy: 0.964519906323185
Desvio 0.02583682959724904
Intervalo [0.913,1.016]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [1.         0.5        0.5        0.97058824 0.5        0.5
 0.5       ]
Média do test_precision_macro: 0.638655462184874
Desvio 0.21937441146565734
Intervalo [0.200,1.077]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-


/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221

Tivemos os seguintes resultados com svm:

- sem *undersampling*:
  - acuracia: 0,973
  - precisão: 0,968
  - recall: 0,970

- com *undersampling*:
  - acuracia: 0,964
  - precisão: 0,638
  - recall: 0.622

#### Aplicando o knn

In [15]:
metricas_ran_k = cross_validate(knn,x_std, y, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_k:
  print(f'-{met}')
  print(f"-- {metricas_ran_k[met]}")
  media = np.mean(metricas_ran_k[met])
  desvio = np.std(metricas_ran_k[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.00293112 0.00722289 0.00404191 0.00309014 0.00319982 0.00297284
 0.00650454]
Média do fit_time: 0.004280464989798409
Desvio 0.0016808396963087602
Intervalo [0.001,0.008]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.01775265 0.02249599 0.02339196 0.01875973 0.0224216  0.0238378
 0.03264213]
Média do score_time: 0.023043121610369
Desvio 0.00447079938598467
Intervalo [0.014,0.032]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.93902439 0.96341463 0.96296296 0.96296296 0.97530864 0.98765432
 0.95061728]
Média do test_accuracy: 0.9631350281756784
Desvio 0.01457760889307831
Intervalo [0.934,0.992]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.92160052 0.97272727 0.96375921 0.97368421 0.98333333 0.99230769
 0.92616192]
Média do test_precision_macro: 0.9619391660269682
Desvio 0.025480573549109344
Intervalo [0.911,1.013]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-

In [16]:
metricas_ran_k_u = cross_validate(knn,x_nm, y_nm, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_k_u:
  print(f'-{met}')
  print(f"-- {metricas_ran_k_u[met]}")
  media = np.mean(metricas_ran_k_u[met])
  desvio = np.std(metricas_ran_k_u[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.00905323 0.00391936 0.00662851 0.00239801 0.00404239 0.00324059
 0.00334501]
Média do fit_time: 0.004661015101841518
Desvio 0.0021701220796458286
Intervalo [0.000,0.009]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.04561925 0.05836225 0.01963687 0.01679158 0.01601577 0.01395822
 0.01569033]
Média do score_time: 0.026582036699567522
Desvio 0.016501609126417776
Intervalo [-0.006,0.060]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [1.         0.95081967 1.         0.96721311 0.88333333 0.91666667
 0.95      ]
Média do test_accuracy: 0.952576112412178
Desvio 0.039290246429388734
Intervalo [0.874,1.031]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [1.         0.5        1.         0.97058824 0.5        0.5
 0.5       ]
Média do test_precision_macro: 0.7100840336134454
Desvio 0.24275389622948385
Intervalo [0.225,1.196]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*

/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221

Tivemos os seguintes resultados com knn:

- sem *undersampling*:
  - acuracia: 0,963
  - precisão: 0,961
  - recall: 0,954

- com *undersampling*:
  - acuracia: 0,952
  - precisão: 0,71
  - recall: 0.687

#### Aplicando o naive bayes

In [17]:
metricas_ran_n = cross_validate(na_ba,x_std, y, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_n:
  print(f'-{met}')
  print(f"-- {metricas_ran_n[met]}")
  media = np.mean(metricas_ran_n[met])
  desvio = np.std(metricas_ran_n[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.00214696 0.00846601 0.0021553  0.00355291 0.00283575 0.00261617
 0.00198817]
Média do fit_time: 0.0033944674900599887
Desvio 0.0021291509302757937
Intervalo [-0.001,0.008]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.02917409 0.01738071 0.0045445  0.00476956 0.00507808 0.00417995
 0.00405383]
Média do score_time: 0.009882961000714983
Desvio 0.009044334634852522
Intervalo [-0.008,0.028]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.90243902 0.86585366 0.96296296 0.92592593 0.98765432 0.95061728
 0.92592593]
Média do test_accuracy: 0.9316255860971309
Desvio 0.037272565554412544
Intervalo [0.857,1.006]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.88076923 0.85718933 0.9673913  0.91666667 0.99152542 0.91298812
 0.8958021 ]
Média do test_precision_macro: 0.9174760242537638
Desvio 0.043829667677608056
Intervalo [0.830,1.005]
-*--*--*--*--*--*--*--*--*--*--*--*--*-

In [18]:
metricas_ran_n_u = cross_validate(na_ba,x_nm, y_nm, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_n_u:
  print(f'-{met}')
  print(f"-- {metricas_ran_n_u[met]}")
  media = np.mean(metricas_ran_n_u[met])
  desvio = np.std(metricas_ran_n_u[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221

-fit_time
-- [0.02772737 0.01573825 0.00871086 0.00249219 0.00355744 0.01354694
 0.01089501]
Média do fit_time: 0.011809723717825753
Desvio 0.00790279135055553
Intervalo [-0.004,0.028]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.01880622 0.02447867 0.01328754 0.00949907 0.016644   0.00754166
 0.01218104]
Média do score_time: 0.014634030205862862
Desvio 0.005383358890724156
Intervalo [0.004,0.025]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.96721311 0.93442623 0.98360656 0.96721311 0.8        0.91666667
 0.93333333]
Média do test_accuracy: 0.9289227166276347
Desvio 0.057000816326885966
Intervalo [0.815,1.043]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.5        0.5        0.5        0.97058824 0.5        0.5
 0.5       ]
Média do test_precision_macro: 0.5672268907563025
Desvio 0.16467157934676827
Intervalo [0.238,0.897]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-

Tivemos os seguintes resultados com naive bayes:

- sem *undersampling*:
  - acuracia: 0,931
  - precisão: 0,917
  - recall: 0,925

- com *undersampling*:
  - acuracia: 0,928
  - precisão: 0,567
  - recall: 0.533

#### Aplicando árvore de decisão

In [19]:
metricas_ran_tree = cross_validate(tree,x_std, y, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_tree:
  print(f'-{met}')
  print(f"-- {metricas_ran_tree[met]}")
  media = np.mean(metricas_ran_tree[met])
  desvio = np.std(metricas_ran_tree[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [0.07474041 0.04207158 0.03279448 0.03922176 0.03589368 0.02030325
 0.0164187 ]
Média do fit_time: 0.037349121911185126
Desvio 0.01762880946024483
Intervalo [0.002,0.073]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.02233624 0.00481319 0.00516081 0.01832843 0.0074544  0.00583529
 0.00744295]
Média do score_time: 0.01019590241568429
Desvio 0.006567899238451153
Intervalo [-0.003,0.023]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.8902439  0.90243902 0.92592593 0.92592593 0.90123457 0.9382716
 0.85185185]
Média do test_accuracy: 0.9051275433389254
Desvio 0.026848462891704606
Intervalo [0.851,0.959]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.87104377 0.90013228 0.92560976 0.91666667 0.87321429 0.89180328
 0.80555556]
Média do test_precision_macro: 0.8834322269855198
Desvio 0.036946088474750866
Intervalo [0.810,0.957]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-

In [20]:
metricas_ran_tree_u = cross_validate(tree,x_nm, y_nm, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_tree_u:
  print(f'-{met}')
  print(f"-- {metricas_ran_tree_u[met]}")
  media = np.mean(metricas_ran_tree_u[met])
  desvio = np.std(metricas_ran_tree_u[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


-fit_time
-- [0.03745961 0.02941394 0.05522633 0.03071642 0.02391505 0.03261828
 0.03380203]
Média do fit_time: 0.03473595210484096
Desvio 0.0092133661521802
Intervalo [0.016,0.053]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.01821566 0.04135323 0.00878    0.00702095 0.00904632 0.00654101
 0.00719523]
Média do score_time: 0.01402177129473005
Desvio 0.011761943475837894
Intervalo [-0.010,0.038]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.93442623 0.8852459  0.93442623 0.93442623 0.9        0.9
 0.95      ]
Média do test_accuracy: 0.9197892271662764
Desvio 0.022462300702731676
Intervalo [0.875,0.965]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.5        0.5        0.5        0.94444444 0.5        0.5
 0.5       ]
Média do test_precision_macro: 0.5634920634920635
Desvio 0.15552315827194782
Intervalo [0.252,0.875]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-


/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Tivemos os seguintes resultados com árvores de decisão:

- sem *undersampling*:
  - acuracia: 0,905
  - precisão: 0,883
  - recall: 0,898

- com *undersampling*:
  - acuracia: 0,929
  - precisão: 0,562
  - recall: 0.531

#### Aplicando regressão logistica

In [23]:
metricas_ran_lg = cross_validate(lg,x_std, y, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_lg:
  print(f'-{met}')
  print(f"-- {metricas_ran_lg[met]}")
  media = np.mean(metricas_ran_lg[met])
  desvio = np.std(metricas_ran_lg[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

-fit_time
-- [6.39604068 0.048985   0.08447194 0.09914184 0.0689528  0.06971669
 0.09780693]
Média do fit_time: 0.9807308401380267
Desvio 2.2108515080257654
Intervalo [-3.441,5.402]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.04685664 0.00571632 0.01188612 0.00407338 0.01199245 0.0118103
 0.00405765]
Média do score_time: 0.013770409992762975
Desvio 0.013930905434481451
Intervalo [-0.014,0.042]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.96341463 0.98780488 0.96296296 0.97530864 0.98765432 0.98765432
 0.98765432]
Média do test_accuracy: 0.9789220114423366
Desvio 0.01079610661969139
Intervalo [0.957,1.001]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.95387205 0.99056604 0.9673913  0.98214286 0.97916667 0.97222222
 0.97727273]
Média do test_precision_macro: 0.9746619813228862
Desvio 0.010869662561705712
Intervalo [0.953,0.996]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*

In [24]:
metricas_ran_lg_u = cross_validate(lg,x_nm, y_nm, cv=kfold, scoring=nome_metricas)

for met in metricas_ran_tree_u:
  print(f'-{met}')
  print(f"-- {metricas_ran_lg_u[met]}")
  media = np.mean(metricas_ran_lg_u[met])
  desvio = np.std(metricas_ran_lg_u[met])
  print(f'Média do {met}: {media}')
  print(f'Desvio {desvio}')
  print(f'Intervalo [{(media-(2*desvio)):.3f},{(media+(2*desvio)):.3f}]')
  print('-*-'*20)

/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


-fit_time
-- [0.13370872 0.08587956 0.24143863 0.15494394 0.06332064 0.0516479
 0.11844873]
Média do fit_time: 0.12134116036551339
Desvio 0.05999396086699831
Intervalo [0.001,0.241]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-score_time
-- [0.00440168 0.01249838 0.00932503 0.00504827 0.00379276 0.00991106
 0.00844407]
Média do score_time: 0.007631608418055943
Desvio 0.0030296600077563437
Intervalo [0.002,0.014]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_accuracy
-- [0.98360656 0.96721311 0.98360656 0.95081967 0.91666667 0.96666667
 1.        ]
Média do test_accuracy: 0.9669398907103826
Desvio 0.025173495663225704
Intervalo [0.917,1.017]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*-
-test_precision_macro
-- [0.5        0.5        0.5        0.95714286 0.5        0.5
 1.        ]
Média do test_precision_macro: 0.636734693877551
Desvio 0.2164997381285851
Intervalo [0.204,1.070]
-*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--*--

/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/gustavo/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


regressão logistica *sem undersampling*:
    - acuracia: 0,978
    - precisão: 0,974
    - recall: 0,977
regressão logistica *com undersampling*:  
    - acuracia: 0,966
    - precisão: 0,636
    - recall: 0,622

podemos ver que:
- aplicando o *undersampling* o resultado saiu pior que não aplicando.
  
Condiderando as médias do *cross validation*
- *regressão logistica* foi o que teve melhor resultado.
- A *árvore de decisão* foi o que teve os priores resultados.
- *svm* foi o que teve a segunda melhor perfomace  com pouca diferença em relação a  *regressão logistica*.
- Todas as métricas estão em torno de 90%.

Agora resta verifica se usando metodos de ensemble e mudança nos parametros se da pra melhorar os resultados.